sysv-pdp11_tape/man/u_man/man1/stty.1
.TH STTY 1
.SH NAME
stty \- set the options for a terminal
.SH SYNOPSIS
.B stty
[
.B \-a
]
[
.B \-g
] [ options ]
.SH DESCRIPTION
.I Stty\^
sets certain terminal \s-1I/O\s+1 options for the device that is
the current standard input;
without arguments, it reports the settings of certain options;
with the
.B \-a
option, it reports all of the option settings;
with the
.B \-g
option, it reports current settings in a form that can be used as an argument
to another
.I stty\^
command.
Detailed information about the modes listed in the first five groups below
may be found in
.IR termio (7)
for asynchronous lines,
or in
.IR stermio (7)
for synchronous lines in the
.I "U\s-1NIX\s+1 System Administrator's Manual" .
Options in the last group are implemented using options in the previous
groups.
Note that many combinations of options make no sense, but no sanity
checking is performed.
The options are
selected from the following:
.PD 1u
.SS Control Modes
.TP 10m
.BR parenb " (" \-parenb )
enable (disable) parity generation and detection.
.TP
.BR parodd " (" \-parodd )
select odd (even) parity.
.TP
.B "cs5 cs6 cs7 cs8"
select character size (see
.IR termio (7)).
.TP
.B 0
hang up phone line immediately.
.TP
.B "50 75 110 134 150 200 300 600 1200 1800 2400 4800 9600 exta extb"
Set terminal baud rate to the number given, if possible.
(All speeds are not supported by all hardware interfaces.)
.TP
.BR hupcl " (" \-hupcl )
hang up (do not hang up) a
.SM DATA-PHONE\*S\*R
data set connection on last close.
.TP
.BR hup " (" \-hup )
same as
.BR hupcl " (" \-hupcl ).
.TP
.BR cstopb " (" \-cstopb )
use two (one) stop bits per character.
.TP
.BR cread " (" \-cread )
enable (disable) the receiver.
.TP
.BR clocal " (" \-clocal )
assume a line without (with) modem control.
.SS Input Modes
.TP 10m
.BR ignbrk " (" \-ignbrk )
ignore (do not ignore) break on input.
.TP
.BR brkint " (" \-brkint )
signal (do not signal) \s-1INTR\s+1 on break.
.TP
.BR ignpar " (" \-ignpar )
ignore (do not ignore) parity errors.
.TP
.BR parmrk " (" \-parmrk )
mark (do not mark) parity errors (see
.IR termio (7)).
.TP
.BR inpck " (" \-inpck )
enable (disable) input parity checking.
.TP
.BR istrip " (" \-istrip )
strip (do not strip) input characters to seven bits.
.TP
.BR inlcr " (" \-inlcr )
map (do not map) \s-1NL\s+1 to \s-1CR\s+1 on input.
.TP
.BR igncr " (" \-igncr )
ignore (do not ignore) \s-1CR\s+1 on input.
.TP
.BR icrnl " (" \-icrnl )
map (do not map) \s-1CR\s+1 to \s-1NL\s+1 on input.
.TP
.BR iuclc " (" \-iuclc )
map (do not map) upper-case alphabetics to lower case on input.
.TP
.BR ixon " (" \-ixon )
enable (disable) \s-1START\s+1/\s-1STOP\s+1
output control.  Output is stopped by sending
an \s-1ASCII DC3\s+1 and started by sending an \s-1ASCII DC1\s+1.
.TP
.BR ixany " (" \-ixany )
allow any character (only \s-1DC1\s+1) to restart output.
.TP
.BR ixoff " (" \-ixoff )
request that the system send (not send) \s-1START/STOP\s+1 characters
when the input queue is nearly empty/full.
.SS Output Modes
.TP 10m
.BR opost " (" \-opost )
post-process output (do not post-process output; ignore all other output modes).
.TP
.BR olcuc " (" \-olcuc )
map (do not map) lower-case alphabetics to upper case on output.
.TP
.BR onlcr " (" \-onlcr )
map (do not map) \s-1NL\s+1 to \s-1CR-NL\s+1 on output.
.TP
.BR ocrnl " (" \-ocrnl )
map (do not map) \s-1CR\s+1 to \s-1NL\s+1 on output.
.TP
.BR onocr " (" \-onocr )
do not (do) output \s-1CR\s+1s at column zero.
.TP
.BR onlret " (" \-onlret )
on the terminal \s-1NL\s+1 performs (does not perform) the \s-1CR\s+1 function.
.TP
.BR ofill " (" \-ofill )
use fill characters (use timing) for delays.
.TP
.BR ofdel " (" \-ofdel )
fill characters are \s-1DEL\s+1s (\s-1NUL\s+1s).
.TP
.B "cr0 cr1 cr2 cr3"
select style of delay for carriage returns (see
.IR termio (7)).
.TP
.B "nl0 nl1"
select style of delay for line-feeds (see
.IR termio (7)).
.TP
.B "tab0 tab1 tab2 tab3"
select style of delay for horizontal tabs (see
.IR termio (7)
or
.IR stermio (7)).
.TP
.B "bs0 bs1"
select style of delay for backspaces (see
.IR termio (7)).
.TP
.B "ff0 ff1"
select style of delay for form-feeds (see
.IR termio (7)).
.TP
.B "vt0 vt1"
select style of delay for vertical tabs (see
.IR termio (7)).
.SS Local Modes
.TP 10m
.BR isig " (" \-isig )
enable (disable) the checking of characters against the special control
characters \s-1INTR\s+1 and \s-1QUIT\s+1.
.TP
.BR icanon " (" \-icanon )
enable (disable) canonical input (\s-1ERASE\s+1 and \s-1KILL\s+1 processing).
.TP
.BR xcase " (" \-xcase )
canonical (unprocessed) upper/lower-case presentation.
.TP
.BR echo " (" \-echo )
echo back (do not echo back) every character typed.
.TP
.BR echoe " (" \-echoe )
echo (do not echo) \s-1ERASE\s+1 character as
a backspace-space-backspace string.  Note: this mode will erase the
\s-1ERASE\s+1ed character on many \s-1CRT\s+1 terminals;
however, it does
.I not\^
keep track of column position and, as a result, may be confusing
on escaped characters, tabs, and backspaces.
.TP
.BR echok " (" \-echok )
echo (do not echo) \s-1NL\s+1 after \s-1KILL\s+1 character.
.TP
.BR lfkc " (" \-lfkc )
the same as
.BR echok " (" \-echok );
obsolete.
.TP
.BR echonl " (" \-echonl )
echo (do not echo) \s-1NL\s+1.
.TP
.BR noflsh " (" \-noflsh )
disable (enable) flush after \s-1INTR\s+1 or \s-1QUIT\s+1.
.TP
.BR stwrap " (" \-stwrap )
disable (enable) truncation of lines longer than 79 characters on a synchronous
line.
.TP
.BR stflush " (" \-stflush )
enable (disable) flush on a synchronous line after every
.IR write (2).
.TP
.BR stappl " (" \-stappl )
use application mode (use line mode) on a synchronous line.
.SS Control Assignments
.TP 10m
.I "control-character c\^"
set
.I control-character\^
to
.IR c ,
where
.I control-character\^
is
.BR erase ", " kill ", " intr ,
.BR quit ", " eof ", " eol ,
.BR ctab ", " min ", or " time
.RB ( ctab
is used with
.BR \-stappl "; see"
.IR stermio (7)),
.RB ( min " and " time
are used with
.BR \-icanon "; see"
.IR termio (7)).
If
.I c\^
is preceded by an (escaped from the shell) caret
.RB ( ^ ),
then the value used is the corresponding \s-1CTRL\s+1 character
(e.g.,
.RB `` ^d ''
is a
.BR \s-1CTRL\s+1-d );
.RB `` ^? ''
is interpreted as \s-1DEL\s+1 and
.RB `` ^\- ''
is interpreted as undefined.
.TP
.BI line " i\^"
set line discipline to
.I i\^
(0 <
.I i\^
< 127
).
.SS Combination Modes
.TP 10m
.BR evenp " or " parity
enable
.BR parenb " and " cs7 .
.TP
.B oddp
enable
.BR parenb ", " cs7 ", and " parodd .
.TP
.BR \-parity ", " \-evenp ", or " \-oddp
disable
.BR parenb ,
and set
.BR cs8 .
.TP
.BR raw " (" \-raw " or " cooked )
enable (disable) raw input and output
(no \s-1ERASE\s+1, \s-1KILL\s+1, \s-1INTR\s+1, \s-1QUIT\s+1,
\s-1EOT\s+1, or output post processing).
.TP
.BR nl " (" \-nl )
unset (set)
.BR icrnl ", " onlcr .
In addition
.B \-nl
unsets
.BR inlcr ", " igncr ", "
.BR ocrnl ", and " onlret .
.TP
.BR lcase " (" \-lcase )
set (unset)
.BR xcase ", " iuclc ", and " olcuc .
.TP
.BR \s-1LCASE\s+1 " (" \-\s-1LCASE\s+1 )
same as
.BR lcase " (" \-lcase ).
.TP
.BR tabs " (" \-tabs " or " tab3 )
preserve (expand to spaces) tabs when printing.
.TP
.B ek
reset \s-1ERASE\s+1 and \s-1KILL\s+1 characters back to normal
.B #
and
.BR @ .
.TP
.B sane
resets all modes to some reasonable values.
.TP
.I term\^
set all modes suitable for the
terminal type
.IR term ,
where
.I term\^
is one of
.BR tty33 ", " tty37 ", " vt05 ", "
.BR tn300 ", " ti700 ", or " tek .
.SH "SEE ALSO"
tabs(1), ioctl(2).
.br
stermio(7), termio(7) in the
.IR "U\s-1NIX\s+1 System Adminstrator's Manual" .
.\"	@(#)stty.1	5.2 of 5/18/82


sysvr2-vax/src/cmd/stty.c
/*	@(#)stty.c	1.6	*/

#include "stdio.h"
#include "sys/types.h"
#include "sys/termio.h"

#include "sys/stermio.h"
#define ASYNC	0
#define SYNC	1
extern char *getenv();

struct
{
	char	*string;
	int	speed;
} speeds[] = {
	"0",	B0,
	"50",	B50,
	"75",	B75,
	"110",	B110,
	"134",	B134,
	"134.5",B134,
	"150",	B150,
	"200",	B200,
	"300",	B300,
	"600",	B600,
	"1200",	B1200,
	"1800",	B1800,
	"2400",	B2400,
	"4800",	B4800,
	"9600",	B9600,
	"exta",	EXTA,
	"extb",	EXTB,
	0,
};
struct mds {
	char	*string;
	int	set;
	int	reset;
};

struct mds cmodes[] = {
	"-parity", CS8, PARENB|CSIZE,
	"-evenp", CS8, PARENB|CSIZE,
	"-oddp", CS8, PARENB|PARODD|CSIZE,
	"parity", PARENB|CS7, PARODD|CSIZE,
	"evenp", PARENB|CS7, PARODD|CSIZE,
	"oddp", PARENB|PARODD|CS7, CSIZE,
	"parenb", PARENB, 0,
	"-parenb", 0, PARENB,
	"parodd", PARODD, 0,
	"-parodd", 0, PARODD,
	"cs8", CS8, CSIZE,
	"cs7", CS7, CSIZE,
	"cs6", CS6, CSIZE,
	"cs5", CS5, CSIZE,
	"cstopb", CSTOPB, 0,
	"-cstopb", 0, CSTOPB,
	"hupcl", HUPCL, 0,
	"hup", HUPCL, 0,
	"-hupcl", 0, HUPCL,
	"-hup", 0, HUPCL,
	"clocal", CLOCAL, 0,
	"-clocal", 0, CLOCAL,
#ifndef pdp11
	"loblk", LOBLK, 0,
	"-loblk", 0, LOBLK,
#endif
	"cread", CREAD, 0,
	"-cread", 0, CREAD,
	"raw", CS8, (CSIZE|PARENB),
	"-raw", (CS7|PARENB), CSIZE,
	"cooked", (CS7|PARENB), CSIZE,
	"sane", (CS7|PARENB|CREAD), (CSIZE|PARODD|CLOCAL),
	0
};

struct mds imodes[] = {
	"ignbrk", IGNBRK, 0,
	"-ignbrk", 0, IGNBRK,
	"brkint", BRKINT, 0,
	"-brkint", 0, BRKINT,
	"ignpar", IGNPAR, 0,
	"-ignpar", 0, IGNPAR,
	"parmrk", PARMRK, 0,
	"-parmrk", 0, PARMRK,
	"inpck", INPCK, 0,
	"-inpck", 0,INPCK,
	"istrip", ISTRIP, 0,
	"-istrip", 0, ISTRIP,
	"inlcr", INLCR, 0,
	"-inlcr", 0, INLCR,
	"igncr", IGNCR, 0,
	"-igncr", 0, IGNCR,
	"icrnl", ICRNL, 0,
	"-icrnl", 0, ICRNL,
	"-nl", ICRNL, (INLCR|IGNCR),
	"nl", 0, ICRNL,
	"iuclc", IUCLC, 0,
	"-iuclc", 0, IUCLC,
	"lcase", IUCLC, 0,
	"-lcase", 0, IUCLC,
	"LCASE", IUCLC, 0,
	"-LCASE", 0, IUCLC,
	"ixon", IXON, 0,
	"-ixon", 0, IXON,
	"ixany", IXANY, 0,
	"-ixany", 0, IXANY,
	"ixoff", IXOFF, 0,
	"-ixoff", 0, IXOFF,
	"raw", 0, -1,
	"-raw", (BRKINT|IGNPAR|ISTRIP|ICRNL|IXON), 0,
	"cooked", (BRKINT|IGNPAR|ISTRIP|ICRNL|IXON), 0,
	"sane", (BRKINT|IGNPAR|ISTRIP|ICRNL|IXON),
		(IGNBRK|PARMRK|INPCK|INLCR|IGNCR|IUCLC|IXOFF),
	0
};

struct mds lmodes[] = {
	"isig", ISIG, 0,
	"-isig", 0, ISIG,
	"icanon", ICANON, 0,
	"-icanon", 0, ICANON,
	"xcase", XCASE, 0,
	"-xcase", 0, XCASE,
	"lcase", XCASE, 0,
	"-lcase", 0, XCASE,
	"LCASE", XCASE, 0,
	"-LCASE", 0, XCASE,
	"echo", ECHO, 0,
	"-echo", 0, ECHO,
	"echoe", ECHOE, 0,
	"-echoe", 0, ECHOE,
	"echok", ECHOK, 0,
	"-echok", 0, ECHOK,
	"lfkc", ECHOK, 0,
	"-lfkc", 0, ECHOK,
	"echonl", ECHONL, 0,
	"-echonl", 0, ECHONL,
	"noflsh", NOFLSH, 0,
	"-noflsh", 0, NOFLSH,
	"raw", 0, (ISIG|ICANON|XCASE),
	"-raw", (ISIG|ICANON), 0,
	"cooked", (ISIG|ICANON), 0,
	"sane", (ISIG|ICANON|ECHO|ECHOK),
		(XCASE|ECHOE|ECHONL|NOFLSH|STFLUSH|STWRAP|STAPPL),
	"stflush", STFLUSH, 0,
	"-stflush", 0, STFLUSH,
	"stwrap", STWRAP, 0,
	"-stwrap", 0, STWRAP,
	"stappl", STAPPL, 0,
	"-stappl", 0, STAPPL,
	0,
};

struct mds omodes[] = {
	"opost", OPOST, 0,
	"-opost", 0, OPOST,
	"olcuc", OLCUC, 0,
	"-olcuc", 0, OLCUC,
	"lcase", OLCUC, 0,
	"-lcase", 0, OLCUC,
	"LCASE", OLCUC, 0,
	"-LCASE", 0, OLCUC,
	"onlcr", ONLCR, 0,
	"-onlcr", 0, ONLCR,
	"-nl", ONLCR, (OCRNL|ONLRET),
	"nl", 0, ONLCR,
	"ocrnl", OCRNL, 0,
	"-ocrnl",0, OCRNL,
	"onocr", ONOCR, 0,
	"-onocr", 0, ONOCR,
	"onlret", ONLRET, 0,
	"-onlret", 0, ONLRET,
	"fill", OFILL, OFDEL,
	"-fill", 0, OFILL|OFDEL,
	"nul-fill", OFILL, OFDEL,
	"del-fill", OFILL|OFDEL, 0,
	"ofill", OFILL, 0,
	"-ofill", 0, OFILL,
	"ofdel", OFDEL, 0,
	"-ofdel", 0, OFDEL,
	"cr0", CR0, CRDLY,
	"cr1", CR1, CRDLY,
	"cr2", CR2, CRDLY,
	"cr3", CR3, CRDLY,
	"tab0", TAB0, TABDLY,
	"tabs", TAB0, TABDLY,
	"tab1", TAB1, TABDLY,
	"tab2", TAB2, TABDLY,
	"tab3", TAB3, TABDLY,
	"-tabs", TAB3, TABDLY,
	"nl0", NL0, NLDLY,
	"nl1", NL1, NLDLY,
	"ff0", FF0, FFDLY,
	"ff1", FF1, FFDLY,
	"vt0", VT0, VTDLY,
	"vt1", VT1, VTDLY,
	"bs0", BS0, BSDLY,
	"bs1", BS1, BSDLY,
	"raw", 0, OPOST,
	"-raw", OPOST, 0,
	"cooked", OPOST, 0,
	"tty33", CR1, (CRDLY|TABDLY|NLDLY|FFDLY|VTDLY|BSDLY),
	"tn300", CR1, (CRDLY|TABDLY|NLDLY|FFDLY|VTDLY|BSDLY),
	"ti700", CR2, (CRDLY|TABDLY|NLDLY|FFDLY|VTDLY|BSDLY),
	"vt05", NL1, (CRDLY|TABDLY|NLDLY|FFDLY|VTDLY|BSDLY),
	"tek", FF1, (CRDLY|TABDLY|NLDLY|FFDLY|VTDLY|BSDLY),
	"tty37", (FF1|VT1|CR2|TAB1|NL1), (NLDLY|CRDLY|TABDLY|BSDLY|VTDLY|FFDLY),
	"sane", (OPOST|ONLCR), (OLCUC|OCRNL|ONOCR|ONLRET|OFILL|OFDEL|
			NLDLY|CRDLY|TABDLY|BSDLY|VTDLY|FFDLY),
	0,
};

char	*arg;
int	match;
char	*STTY="stty: ";
char	*USAGE="usage: stty [-ag] [modes]\n";
int	pitt = 0;
struct termio cb;
struct stio stio;
int term;

main(argc, argv)
char	*argv[];
{
	register i;
	char *ptr;

	if(ioctl(0, STGET, &stio) == -1) {
		term = ASYNC;
		if(ioctl(0, TCGETA, &cb) == -1) {
			perror(STTY);
			exit(2);
		}
	}
	else {
		term = SYNC;
		cb.c_cc[7] = stio.tab;
		cb.c_lflag = stio.lmode;
		cb.c_oflag = stio.omode;
		cb.c_iflag = stio.imode;
	}

	if (argc == 1) {
		prmodes();
		exit(0);
	}
	if ((argc == 2) && (argv[1][0] == '-') && (argv[1][2] == '\0'))
	switch(argv[1][1]) {
		case 'a':
			pramodes();
			exit(0);
		case 'g':
			prencode();
			exit(0);
		default:
			fprintf(stderr, "%s", USAGE);
			exit(2);
	}
	while(--argc > 0) {

		arg = *++argv;
		match = 0;
		if (term == ASYNC) {
			if (eq("erase") && --argc)
				cb.c_cc[VERASE] = gct(*++argv);
			else if (eq("intr") && --argc)
				cb.c_cc[VINTR] = gct(*++argv);
			else if (eq("quit") && --argc)
				cb.c_cc[VQUIT] = gct(*++argv);
			else if (eq("eof") && --argc)
				cb.c_cc[VEOF] = gct(*++argv);
			else if (eq("min") && --argc)
				cb.c_cc[VMIN] = gct(*++argv);
			else if (eq("eol") && --argc)
				cb.c_cc[VEOL] = gct(*++argv);
			else if (eq("time") && --argc)
				cb.c_cc[VTIME] = gct(*++argv);
			else if (eq("kill") && --argc)
				cb.c_cc[VKILL] = gct(*++argv);
#ifndef pdp11
			else if (eq("swtch") && --argc)
				cb.c_cc[VSWTCH] = gct(*++argv);
#endif
			else if (eq("ek")) {
				cb.c_cc[VERASE] = CERASE;
				cb.c_cc[VKILL] = CKILL;
			}
			else if (eq("line") && --argc)
				cb.c_line = atoi(*++argv);
			else if (eq("raw")) {
				cb.c_cc[VMIN] = 1;
				cb.c_cc[VTIME] = 1;
			}
			else if (eq("-raw") | eq("cooked")) {
				cb.c_cc[VEOF] = CEOF;
				cb.c_cc[VEOL] = CNUL;
			}
			else if(eq("sane")) {
				cb.c_cc[VERASE] = CERASE;
				cb.c_cc[VKILL] = CKILL;
				cb.c_cc[VQUIT] = CQUIT;
				cb.c_cc[VINTR] = CINTR;
				cb.c_cc[VEOF] = CEOF;
				cb.c_cc[VEOL] = CNUL;
#ifndef pdp11
				cb.c_cc[VSWTCH] = CNUL;
#endif
			}
			for(i=0; speeds[i].string; i++)
				if(eq(speeds[i].string)) {
					cb.c_cflag &= ~CBAUD;
					cb.c_cflag |= speeds[i].speed&CBAUD;
				}
		}
		if (term == SYNC && eq("ctab") && --argc)
			cb.c_cc[7] = gct(*++argv);
		for(i=0; imodes[i].string; i++)
			if(eq(imodes[i].string)) {
				cb.c_iflag &= ~imodes[i].reset;
				cb.c_iflag |= imodes[i].set;
			}
		for(i=0; omodes[i].string; i++)
			if(eq(omodes[i].string)) {
				cb.c_oflag &= ~omodes[i].reset;
				cb.c_oflag |= omodes[i].set;
			}
		if(term == SYNC && eq("sane"))
			cb.c_oflag |= TAB3;
		for(i=0; cmodes[i].string; i++)
			if(eq(cmodes[i].string)) {
				cb.c_cflag &= ~cmodes[i].reset;
				cb.c_cflag |= cmodes[i].set;
			}
		for(i=0; lmodes[i].string; i++)
			if(eq(lmodes[i].string)) {
				cb.c_lflag &= ~lmodes[i].reset;
				cb.c_lflag |= lmodes[i].set;
			}
		if(!match)
			if(!encode(arg)) {
				fprintf(stderr, "unknown mode: %s\n", arg);
				exit(2);
			}
	}
	if (term == ASYNC) {
		if(ioctl(0, TCSETAW, &cb) == -1) {
			perror(STTY);
			exit(2);
		}
	} else {
		stio.imode = cb.c_iflag;
		stio.omode = cb.c_oflag;
		stio.lmode = cb.c_lflag;
		stio.tab = cb.c_cc[7];
		if (ioctl(0, STSET, &stio) == -1) {
			perror (STTY);
			exit(2);
		}
	}
	exit(0);
}

eq(string)
char *string;
{
	register i;

	if(!arg)
		return(0);
	i = 0;
loop:
	if(arg[i] != string[i])
		return(0);
	if(arg[i++] != '\0')
		goto loop;
	match++;
	return(1);
}

prmodes()
{
	register m;

	if (term == SYNC) {
		m = stio.imode;
		if (m & IUCLC) printf ("iuclc ");
		else printf ("-iuclc ");
		m = stio.omode;
		if (m & OLCUC) printf ("olcuc ");
		else printf ("-olcuc ");
		if (m & TAB3) printf ("tab3 ");
		m = stio.lmode;
		if (m & XCASE) printf ("xcase ");
		else printf ("-xcase ");
		if (m & STFLUSH) printf ("stflush ");
		else printf ("-stflush ");
		if (m & STWRAP) printf ("stwrap ");
		else printf ("-stwrap ");
		if (m & STAPPL) printf ("stappl ");
		else printf ("-stappl ");
		printf ("\n");
	}
	if (term == ASYNC) {
		m = cb.c_cflag;
		prspeed("speed ", m&CBAUD);
		if (m&PARENB)
			if (m&PARODD)
				printf("oddp ");
			else
				printf("evenp ");
		else
			printf("-parity ");
		if(((m&PARENB) && !(m&CS7)) || (!(m&PARENB) && !(m&CS8)))
			printf("cs%c ",'5'+(m&CSIZE)/CS6);
		if (m&CSTOPB)
			printf("cstopb ");
		if (m&HUPCL)
			printf("hupcl ");
		if (!(m&CREAD))
			printf("cread ");
		if (m&CLOCAL)
			printf("clocal ");
#ifndef pdp11
		if (m&LOBLK)
			printf("loblk");
#endif
		printf("\n");
		if(cb.c_line != 0)
			printf("line = %d; ", cb.c_line);
		if(cb.c_cc[VINTR] != CINTR)
			pit(cb.c_cc[VINTR], "intr", "; ");
		if(cb.c_cc[VQUIT] != CQUIT)
			pit(cb.c_cc[VQUIT], "quit", "; ");
		if(cb.c_cc[VERASE] != CERASE)
			pit(cb.c_cc[VERASE], "erase", "; ");
		if(cb.c_cc[VKILL] != CKILL)
			pit(cb.c_cc[VKILL], "kill", "; ");
		if(cb.c_cc[VEOF] != CEOF)
			pit(cb.c_cc[VEOF], "eof", "; ");
		if(cb.c_cc[VEOL] != CNUL)
			pit(cb.c_cc[VEOL], "eol", "; ");
#ifndef pdp11
		if(cb.c_cc[VSWTCH] != CSWTCH)
			pit(cb.c_cc[VSWTCH], "swtch", "; ");
#endif
		if(pitt) printf("\n");
		m = cb.c_iflag;
		if (m&IGNBRK)
			printf("ignbrk ");
		else if (m&BRKINT)
			printf("brkint ");
		if (!(m&INPCK))
			printf("-inpck ");
		else if (m&IGNPAR)
			printf("ignpar ");
		if (m&PARMRK)
			printf("parmrk ");
		if (!(m&ISTRIP))
			printf("-istrip ");
		if (m&INLCR)
			printf("inlcr ");
		if (m&IGNCR)
			printf("igncr ");
		if (m&ICRNL)
			printf("icrnl ");
		if (m&IUCLC)
			printf("iuclc ");
		if (!(m&IXON))
			printf("-ixon ");
		else if (!(m&IXANY))
			printf("-ixany ");
		if (m&IXOFF)
			printf("ixoff ");
		m = cb.c_oflag;
		if (!(m&OPOST))
			printf("-opost ");
		else {
		if (m&OLCUC)
			printf("olcuc ");
		if (m&ONLCR)
			printf("onlcr ");
		if (m&OCRNL)
			printf("ocrnl ");
		if (m&ONOCR)
			printf("onocr ");
		if (m&ONLRET)
			printf("onlret ");
		if (m&OFILL)
			if (m&OFDEL)
				printf("del-fill ");
			else
				printf("nul-fill ");
		delay((m&CRDLY)/CR1, "cr");
		delay((m&NLDLY)/NL1, "nl");
		delay((m&TABDLY)/TAB1, "tab");
		delay((m&BSDLY)/BS1, "bs");
		delay((m&VTDLY)/VT1, "vt");
		delay((m&FFDLY)/FF1, "ff");
		}
		printf("\n");
		m = cb.c_lflag;
		if (!(m&ISIG))
			printf("-isig ");
		if (!(m&ICANON))
			printf("-icanon ");
		if (m&XCASE)
			printf("xcase ");
		printf("-echo "+((m&ECHO)!=0));
		printf("-echoe "+((m&ECHOE)!=0));
		printf("-echok "+((m&ECHOK)!=0));
		if (m&ECHONL)
			printf("echonl ");
		if (m&NOFLSH)
			printf("noflsh ");
		printf("\n");
	}
}

pramodes()
{
	register m;

	if(term == ASYNC) {
		prspeed("speed ", cb.c_cflag&CBAUD);
		printf("line = %d; ", cb.c_line);
		pit(cb.c_cc[VINTR], "intr", "; ");
		pit(cb.c_cc[VQUIT], "quit", "; ");
		pit(cb.c_cc[VERASE], "erase", "; ");
		pit(cb.c_cc[VKILL], "kill", "; ");
		pit(cb.c_cc[VEOF], "eof", "; ");
		pit(cb.c_cc[VEOL], "eol", "; ");
#ifndef pdp11
		pit(cb.c_cc[VSWTCH], "swtch", "\n");
#endif
	} else
		pit(stio.tab, "ctab", "\n");
	m = cb.c_cflag;
	printf("-parenb "+((m&PARENB)!=0));
	printf("-parodd "+((m&PARODD)!=0));
	printf("cs%c ",'5'+(m&CSIZE)/CS6);
	printf("-cstopb "+((m&CSTOPB)!=0));
	printf("-hupcl "+((m&HUPCL)!=0));
	printf("-cread "+((m&CREAD)!=0));
	printf("-clocal "+((m&CLOCAL)!=0));

#ifndef pdp11
	printf("-loblk "+((m&LOBLK)!=0));
#endif

	printf("\n");
	m = cb.c_iflag;
	printf("-ignbrk "+((m&IGNBRK)!=0));
	printf("-brkint "+((m&BRKINT)!=0));
	printf("-ignpar "+((m&IGNPAR)!=0));
	printf("-parmrk "+((m&PARMRK)!=0));
	printf("-inpck "+((m&INPCK)!=0));
	printf("-istrip "+((m&ISTRIP)!=0));
	printf("-inlcr "+((m&INLCR)!=0));
	printf("-igncr "+((m&IGNCR)!=0));
	printf("-icrnl "+((m&ICRNL)!=0));
	printf("-iuclc "+((m&IUCLC)!=0));
	printf("\n");
	printf("-ixon "+((m&IXON)!=0));
	printf("-ixany "+((m&IXANY)!=0));
	printf("-ixoff "+((m&IXOFF)!=0));
	printf("\n");
	m = cb.c_lflag;
	printf("-isig "+((m&ISIG)!=0));
	printf("-icanon "+((m&ICANON)!=0));
	printf("-xcase "+((m&XCASE)!=0));
	printf("-echo "+((m&ECHO)!=0));
	printf("-echoe "+((m&ECHOE)!=0));
	printf("-echok "+((m&ECHOK)!=0));
	printf("-echonl "+((m&ECHONL)!=0));
	printf("-noflsh "+((m&NOFLSH)!=0));
	if(term == SYNC) {
		printf("-stflush "+((m&STFLUSH)!=0));
		printf("-stwrap "+((m&STWRAP)!=0));
		printf("-stappl "+((m&STAPPL)!=0));
	}
	printf("\n");
	m = cb.c_oflag;
	printf("-opost "+((m&OPOST)!=0));
	printf("-olcuc "+((m&OLCUC)!=0));
	printf("-onlcr "+((m&ONLCR)!=0));
	printf("-ocrnl "+((m&OCRNL)!=0));
	printf("-onocr "+((m&ONOCR)!=0));
	printf("-onlret "+((m&ONLRET)!=0));
	printf("-ofill "+((m&OFILL)!=0));
	printf("-ofdel "+((m&OFDEL)!=0));
	delay((m&CRDLY)/CR1, "cr");
	delay((m&NLDLY)/NL1, "nl");
	delay((m&TABDLY)/TAB1, "tab");
	delay((m&BSDLY)/BS1, "bs");
	delay((m&VTDLY)/VT1, "vt");
	delay((m&FFDLY)/FF1, "ff");
	printf("\n");
}

gct(cp)
register char *cp;
{
	register c;

	c = *cp++;
	if (c == '^') {
		c = *cp;
		if (c == '?')
			c = 0177;
		else if (c == '-')
			c = 0377;
		else
			c &= 037;
	}
	return(c);
}

pit(what, itsname, sep)
	unsigned char what;
	char *itsname, *sep;
{

	pitt++;
	printf("%s", itsname);
	if (what == 0377) {
		printf(" <undef>%s", sep);
		return;
	}
	printf(" = ");
	if (what & 0200) {
		printf("-");
		what &= ~ 0200;
	}
	if (what == 0177) {
		printf("DEL%s", sep);
		return;
	} else if (what < ' ') {
		printf("^");
		what += '`';
	}
	printf("%c%s", what, sep);
}

delay(m, s)
char *s;
{
	if(m)
		printf("%s%d ", s, m);
}

int	speed[] = {
	0,50,75,110,134,150,200,300,600,1200,1800,2400,4800,9600,0,0
};

prspeed(c, s)
char *c;
{

	printf("%s%d baud; ", c, speed[s]);
}

prencode()
{
	printf("%x:%x:%x:%x:%x:%x:%x:%x:%x:%x:%x:%x\n",
	cb.c_iflag,cb.c_oflag,cb.c_cflag,cb.c_lflag,cb.c_cc[0],
	cb.c_cc[1],cb.c_cc[2],cb.c_cc[3],cb.c_cc[4],cb.c_cc[5],
	cb.c_cc[6],cb.c_cc[7]);
}

encode(arg)
char *arg;
{
	int grab[12], i;
	i = sscanf(arg, "%x:%x:%x:%x:%x:%x:%x:%x:%x:%x:%x:%x",
	&grab[0],&grab[1],&grab[2],&grab[3],&grab[4],&grab[5],&grab[6],
	&grab[7],&grab[8],&grab[9],&grab[10],&grab[11]);

	if(i != 12) return(0);

	cb.c_iflag = (ushort) grab[0];
	cb.c_oflag = (ushort) grab[1];
	cb.c_cflag = (ushort) grab[2];
	cb.c_lflag = (ushort) grab[3];

	for(i=0; i<8; i++)
		cb.c_cc[i] = (char) grab[i+4];
	return(1);
}
